src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,507 @@
+     */
+    private int currentSlot = 0;
+
+    /**
+     * Amount of stacks the worker already kept in the current gathering process.
+     */
+    private List<ItemStorage> alreadyKept = new ArrayList<>();
+
+    /**
      * Initialize the deliveryman and add all his tasks.
      *
      * @param deliveryman the job he has.
      */
     public EntityAIWorkDeliveryman(@NotNull final JobDeliveryman deliveryman)
     {
         super(deliveryman);
+        super.registerTargets(
+                /**
+                 * Check if tasks should be executed.
+                 */
+                new AITarget(this::checkIfExecute, IDLE),
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, this::checkWareHouse),
+                new AITarget(PREPARE_DELIVERY, this::prepareDelivery),
+                new AITarget(GATHER_IN_WAREHOUSE, this::gatherItemsFromWareHouse),
+                new AITarget(DELIVERY, this::deliver),
+                new AITarget(GATHERING, this::gather),
+                new AITarget(DUMPING, this::dump)
+
+        );
+        worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+        worker.setCanPickUpLoot(true);
+    }
+
+    /**
+     * Gather items from a random hut which the hut doesn't need.
+     *
+     * @return the next state to go to.
+     */
+    public AIState gather()
+    {
+        if (gatherTarget == null)
+        {
+            gatherTarget = getRandomBuilding();
+        }
+
+        if (gatherTarget == null)
+        {
+            return START_WORKING;
+        }
+
+        if (!worker.isWorkerAtSiteWithMove(gatherTarget, MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return GATHERING;
+        }
+
+        final Colony colony = getOwnBuilding().getColony();
+        if (colony != null)
+        {
+            final AbstractBuilding building = colony.getBuilding(gatherTarget);
+            if (building == null)
+            {
+                gatherTarget = null;
+                return START_WORKING;
+            }
+            if (gatherFromBuilding(building) || cannotHoldMoreItems())
+            {
+                this.alreadyKept = new ArrayList<>();
+                this.currentSlot = 0;
+                return DUMPING;
+            }
+            currentSlot++;
+            return GATHERING;
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Check if the worker can hold that much items.
+     * It depends on his building level.
+     * Level 1: 1 stack Level 2: 2 stacks, 4 stacks, 8, unlimited.
+     * That's 2^buildingLevel-1.
+     */
+    private boolean cannotHoldMoreItems()
+    {
+        if (getOwnBuilding().getBuildingLevel() >= getOwnBuilding().getMaxBuildingLevel())
+        {
+            return false;
+        }
+        return InventoryUtils.getAmountOfStacks(worker.getInventoryCitizen()) >= Math.pow(2, getOwnBuilding().getBuildingLevel() - 1.0D);
```

Comment:
Can use bit shift there instead of the Math.pow, should be much better than use float point numbers